### ISSUE TITLE
Unpacking an iterable in a list comprehension leads to the type inferred as list[Any]

### ISSUE THREAD

**Issue by hvenev** [python/mypy#15747]
**Bug Report**

When a value of a user-defined class with `__iter__` is converted into a list using `[*value]`, the type of the list is inferred as `list[Any]`.

**To Reproduce**

```python
import typing

class Spam:
    def __iter__(self, /) -> typing.Iterator[int]:
        yield 1

a = Spam()

# list[int]
reveal_type(list(a))

# list[int]
reveal_type([i for i in a])

# list[int]
reveal_type([*(i for i in a)])

# list[int]
reveal_type([*a.__iter__()])

# list[Any] ???
reveal_type([*a])

b, = a
# int
reveal_type(b)
```

**Expected Behavior**

The type of `[*a]` is `list[int]`

**Actual Behavior**
```
example.py:22: note: Revealed type is "builtins.list[Any]"
```

**Your Environment**

- Mypy version used: 1.4.1
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files): ` `
- Python version used: `3.11.4`

**Comment by kamilturek** [comment]
I'll give it a try.

**Comment by JaylenLuc** [comment]
Are you still working on it? I can give it a try as well. Any insights to the issue?

**Comment by JaylenLuc** [comment]
working on it. made progress, ill get back to you with my findings

**Comment by JaylenLuc** [comment]
im trying to figure out how the * unpacking unary operator works. is [*a.__iter__()] the same as reveal_type([*a])?

**Comment by hauntsaninja** [comment]
I didn't understand that comment, but my guess is you'd need to trace the code into argmap

**Comment by JaylenLuc** [comment]
> I didn't understand that comment, but my guess is you'd need to trace the code into argmap.py
thanks, working on it

**Comment by JaylenLuc** [comment]
The function infer_arg_types_in_context() returns List[Type] which turns out to be [testing.Spam] instead of typing.Iterator[builtins.int]. testing.Spam.args is of length 0. MyPy thinks it has no arguments. This is on line 1985

https://github.com/python/mypy/blob/8b73cc22c6a251682f777b104677fa0e1ed5fd67/mypy/checkexpr.py#L1984-L2008

MyPy reaches argmap.py where it then checks whether testing.Spam has any arguments. It sees that it doesnt on line 192.
This results in MyPy returning AnyType(TypeOfAny.from_error) instead of returning the return value of map_instance_to_supertype(). 

https://github.com/python/mypy/blob/8b73cc22c6a251682f777b104677fa0e1ed5fd67/mypy/argmap.py#L192-L217

ISSUE: infer_arg_types_in_context() in checkexpr.py on line 1985 returns the wrong type of [*a]. It is not typing.Iterator[builtins.int] but instead returns a fallacious type of testing.Spam. MyPy then thinks testing.Spam has no arguments and does not unpack them.

How do you all reckon?

**Comment by JaylenLuc** [comment]
in checkexpr.py, the function infer_arg_types_in_context() on line 1985 returns [testing.Spam] when it should return typing.Iterator[builtins.int].

https://github.com/python/mypy/blob/8b73cc22c6a251682f777b104677fa0e1ed5fd67/mypy/checkexpr.py#L1984-L2008

This testing.Spam object is an Instance object where testing.Spam.args length is 0. This means that MyPy thinks testing.Spam has no arguments when it checks in argmap.py line 192 and returns AnyType(TypeOfAny.from_error) on line 217 instead of the result from map_instance_to_supertype().

https://github.com/python/mypy/blob/8b73cc22c6a251682f777b104677fa0e1ed5fd67/mypy/argmap.py#L191-L219

The main issue is that [*a] cannot be unpacked and thus the types of what the Spam object yields is not unpacked.


how do you all reckon?

**Comment by hauntsaninja** [comment]
Yeah, I'm guessing that code dates back before structural types and Protocols. Probably want to swap out the stuff in argmap for some code from checker that analyses the return type of `__iter__`, like `analyze_iterable_item_type`. Might need to thread the checker through to make this work

**Comment by JaylenLuc** [comment]
submitted a PR. if you check the comments Im having issues with dmypy_server.py which is strange since I dont know how my changes are affecting it.

**Comment by huix11** [comment]
3

**Comment by flaeppe** [comment]
I think #14470 reports the same thing as well

**Comment by ilevkivskyi** [comment]
Another case when this happens is when a start argument has union type:
```python
xs: list[str] | list[int]
reveal_type([*xs])  # list[Any]
```
